Sources/NIOHPACK/HPACKHeader.swift

```diff
@@ -276,42 +265,159 @@ public struct HPACKHeaders: ExpressibleByDictionaryLiteral, Sendable {
     /// - Returns: A list of the values for that header field name.
     @inlinable
     public subscript(canonicalForm name: String) -> [String] {
-        let result = self[name]
-        guard result.count > 0 else {
+        return self._values(forHeader: name, canonicalForm: true)
+    }
+
+    /// Return a sequence of header values for the header field named `name`.
+    ///
+    /// See also ``subscript(name:)`` and ``subscript(canonicalForm:)``.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header field whose values should be iterated.
+    ///   - canonicalForm: Whether the values should be decomposed into their "canonical form". That
+    ///     is, splitting them on commas as extensively as possible such that multiple values
+    ///     received on the one line are returned as separate entries.
+    /// - Note: Enabling `canonicalForm`  for the 'set-cookie' header has no efffect.
+    /// - Returns: A sequence of header values for `name`.
+    public func values(forHeader name: String, canonicalForm: Bool = false) -> Values {
+        if canonicalForm, name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
+            // Not safe to split 'set-cookie'.
+            return Values(headers: self, name: name, canonicalize: false)
+        } else {
+            return Values(headers: self, name: name, canonicalize: canonicalForm)
+        }
+    }
+
+    /// Special internal function for use by tests.
+    internal subscript(position: Int) -> (String, String) {
+        precondition(position < self.headers.endIndex, "Position \(position) is beyond bounds of \(self.headers.endIndex)")
+        let header = self.headers[position]
+        return (header.name, header.value)
+    }
+
+    @inlinable
+    internal func _values(forHeader name: String, canonicalForm: Bool) -> [String] {
+        if self.headers.isEmpty {
             return []
         }
 
-        // It's not safe to split Set-Cookie on comma.
-        if name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
-            return result
+        // Estimate the size of the array at the cost of a second scan through the header table.
+        // This allows us to avoid the incremental allocations we would otherwise have without
+        // reserving capacity in the array.
+        let count = self._countValues(forHeader: name)
```

Comment:
This seems somewhat needlessly expensive: is there any reason not to just overcommit the array? We could default this size to `self.count`, for example.

Reply:
I wasn't sure which way to lean with the expense: counting values means we don't allocate in the case where there are no values for the given header (this is the behaviour we have now, at least for `subscript(name:)`). Allocating `self.count` also seems excessive. Don't mind which way we go here.

Reply:
Oh interesting, good spot. If we abandon using `append(contentsOf:)` and instead use a `for in` loop then we can lazily reserve the capacity.

Reply:
We can, but then we get the intermediate (re)allocations which is also a bit sad... 🤷‍♂️

Reply:
Oh, or do you mean just the first iteration?

Reply:
Just the first iteration.